#! /bin/sh /usr/share/dpatch/dpatch-run
## 02_sector_size_error.dpatch by Rob Walker <rob@tenfoot.org.uk>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix a sector size error when file is given as argument
## DP: instead of a block device.

@DPATCH@
--- cryptsetup-1.0.4+svn16/lib/utils.c
+++ cryptsetup-1.0.4+svn16/lib/utils.c
@@ -151,8 +151,10 @@
 static int sector_size(int fd) 
 {
 	int bsize;
-	ioctl(fd,BLKSSZGET, &bsize);
-	return bsize;
+	if (ioctl(fd,BLKSSZGET, &bsize) < 0)
+		return -EINVAL;
+	else
+		return bsize;
 }
 
 int sector_size_for_device(const char *device)
@@ -173,6 +175,9 @@
 	int r;
 	int hangover; int solid; int bsize = sector_size(fd);
 
+	if (bsize < 0)
+		return bsize;
+
 	hangover = count % bsize;
 	solid = count - hangover;
 
@@ -211,6 +216,9 @@
 	int step;
 	int bsize = sector_size(fd);
 
+	if (bsize < 0)
+		return bsize;
+
 	padbuf = aligned_malloc(&padbuf_base, bsize, bsize);
 	if(padbuf == NULL) return -ENOMEM;
 
@@ -242,6 +250,9 @@
 	int frontHang = offset % bsize;
 	int r;
 
+	if (bsize < 0)
+		return bsize;
+
 	lseek(fd, offset - frontHang, SEEK_SET);
 	if(offset % bsize) {
 		int innerCount = count<bsize?count:bsize;
